### ISSUE TITLE
Query fails when returning more than Int32Max rows on v1.0.0

### ISSUE THREAD

**Issue by akamor** [mysql-net/MySqlConnector#862]
We believe we are seeing failures when issuing:

SELECT
     *
FROM
    <table_name> 

on tables with more than 2^31 -1 rows, which is the max value of a signed 32bit integer.  We are working on a minimal repro, however, its failed multiple times and consistently on various tables that have more than int32max rows and the failure always occurs right at 2^31-1 rows and hence we think this is likely a bug in the driver.

Here is the stack trace:

```
System.AggregateException: One or more errors occurred. (One or more errors occurred. (Packet received out-of-order. Expected -255; got 1.))
 ---> System.AggregateException: One or more errors occurred. (Packet received out-of-order. Expected -255; got 1.)
 ---> MySqlConnector.MySqlProtocolException: Packet received out-of-order. Expected -255; got 1.
   at MySqlConnector.Protocol.Serialization.ProtocolUtility.DoReadPayloadAsync(BufferedByteReader bufferedByteReader, IByteHandler byteHandler, Func`1 getNextSequenceNumber, ArraySegmentHolder`1 previousPayloads, ProtocolErrorBehavior protocolErrorBehavior, IOBehavior ioBehavior) in /_/src/MySqlConnector/Protocol/Serialization/ProtocolUtility.cs:line 469
   at MySqlConnector.Protocol.Serialization.StandardPayloadHandler.ReadPayloadAsync(ArraySegmentHolder`1 cache, ProtocolErrorBehavior protocolErrorBehavior, IOBehavior ioBehavior) in /_/src/MySqlConnector/Protocol/Serialization/StandardPayloadHandler.cs:line 42
   at MySqlConnector.Core.ServerSession.ReceiveReplyAsync(IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ServerSession.cs:line 781
--- End of stack trace from previous location where exception was thrown ---
   at MySqlConnector.Core.ServerSession.ReceiveReplyAsyncAwaited(ValueTask`1 task) in /_/src/MySqlConnector/Core/ServerSession.cs:line 793
   at MySqlConnector.Core.ResultSet.<ScanRowAsync>g__ScanRowAsyncAwaited|9_0(ResultSet this_, Task`1 payloadTask, Row row_, CancellationToken token) in /_/src/MySqlConnector/Core/ResultSet.cs:line 244
   at MySqlConnector.Core.ResultSet.ReadAsync(IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ResultSet.cs:line 215
   at MySqlConnector.Core.ResultSet.Read() in /_/src/MySqlConnector/Core/ResultSet.cs:line 191
```

**Comment by bgrainger** [comment]
I don't have a test environment with two billion rows immediately available to me 😀 but I would guess the problem is signed arithmetic on this line: https://github.com/mysql-net/MySqlConnector/blob/62bce553b70fc51609492a1b1449a079085a7966/src/MySqlConnector/Protocol/Serialization/StandardPayloadHandler.cs#L12

**Comment by akamor** [comment]
Yes, that was our conclusion as well.  We aren't familiar with this code base and actually had trouble building it locally but do you think changing m_sequenceNumber (Line 50) to a long would be sufficient here or is there a lot of other changes that would be necessary?

If you could provide some guidance on where a fix would best be placed we can likely craft a PR.

**Comment by bgrainger** [comment]
> actually had trouble building it locally

The `master` branch is currently using some C# 9 features so you will need a .NET 5 Preview SDK to build it currently. (The `v0.x` branch should only need the .NET Core 3.1 SDK.)

I think the best change would be changing `getNextSequenceNumber` (and all supporting code) to a `Func<byte>` and letting regular integer wraparound do the right thing. https://github.com/mysql-net/MySqlConnector/blob/62bce553b70fc51609492a1b1449a079085a7966/src/MySqlConnector/Protocol/Serialization/ProtocolUtility.cs#L404

**Comment by akamor** [comment]
Can you explain a bit the meaning and intent of m_sequenceNumber?

Currently, it looks like it gets incremented once per row, but based on your above comment the driver doesn't actually care how many rows its returned since you state that its OK for it to simply wrap around from 255.

**Comment by akamor** [comment]
FYI, just confirmed our hunch that the sequenceNumber is the issue.  I pulled version 0.69.8 and modified it's type to be a smallint then ran against a table with ~100k rows.  Error occured at row 32768 which is the max value of a smallint.

I'm a bit confused about v1.0.0 and v0.69.8.  I'm going to go ahead and make a PR for this issue.  Should I base it off of master or off of some other branch which somehow corresponds to v0.69?

**Comment by bgrainger** [comment]
It's the [sequence ID](https://dev.mysql.com/doc/internals/en/sequence-id.html) from the [packet header](https://dev.mysql.com/doc/internals/en/mysql-packet.html).

**Comment by akamor** [comment]
Awesome.  I'll get started now.

**Comment by bgrainger** [comment]
I don't want to discourage you from contributing to the project (especially since this is something you can test) but this is also something I should be able to fix quickly (if that works better for you).

**Comment by akamor** [comment]
I'd like to fix it.  I have the fix ready to go and a proper commit message.  I created a new branch, commited, but I seem to not have access to push the branch.  Any thoughts?

**Comment by akamor** [comment]
Ok, I figured it out.  Created a fork...


FYI, going the path of changing the signature for Func<int> to Func<byte> was problematic because Line 42 of CompressedPayloadHandler.cs uses the Func: () => -1 which won't work for a Func<byte>.  It does this so the next time the incrementer gets called it is reset to 0 which is what the spec requires.  Seemed easier to just do what I did in the PR which is to change the m_sequenceNumber to a byte and let the casting be taken care of for us.
